### PR TITLE
Test utils for new spec test format

### LINKED ISSUE #359: Generate test cases from directory path
- next release of spec tests will split each test case into separate file
- if library is given path to directory it should create test case ("it") for each yml file in directory
- test cases should be generated without loading yaml files (traverse directory and generate cases from file names)
- load test file inside of "it" method callback

### LINKED ISSUE #360: Add simple test case to verify library is working
- create small yaml with structure like spec test files
- assert that library call all required functions and assertions

### PR DESCRIPTION
**Pull request by mpetrunic**
- method to generate test cases from new spec test format
- test to ensure library is working

resolves #359
resolves #360

### CODE REVIEW

**Comment by mpetrunic** [comment]
I need to merge master to fix issues but it's ready otherwise

**Comment by GregTheGreek** [comment]
This is still failing the ci

**Comment by mpetrunic** [comment]
> This is still failing the ci

This is new PR, wanted some input before merging master.

**Comment by mpetrunic** [comment]
I mean, PR build is with merged master so that's passing

File: packages/eth2.0-spec-test-util/src/single.ts [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    }
+    if(options.shouldError && options.shouldError(testCase)) {
+      expect(testFunction.bind(null, testCase)).to.throw;
+    } else {
+      const profileId = `${name}-${Date.now()}.profile`;
+      if (process.env.GEN_PROFILE_DIR) {
+        profiler.startProfiling(profileId);
+      }
+      const result = testFunction(testCase);
+      if (process.env.GEN_PROFILE_DIR) {
```
Comment (GregTheGreek) [inline packages/eth2.0-spec-test-util/src/single.ts]: could this not be combined to into the top `if` statement?

**Review by GregTheGreek — CHANGES_REQUESTED** [review]

File: packages/eth2.0-spec-test-util/src/single.ts (hunk 1 above)
Comment (mpetrunic) [reply packages/eth2.0-spec-test-util/src/single.ts]: hmm not really, we would never execute tesFunction

File: packages/eth2.0-spec-test-util/src/single.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+  shouldSkip: () => false,
+  expectFunc: (testCase, expected, actual) => expect(actual).to.be.deep.equal(expected),
+  timeout: 4000
+};
+
+export function describeDirectorySpecTest<TestCase, Result>(
+  name: string,
+  testCaseDirectoryPath: string,
+  testFunction: (testCase: TestCase) => Result,
+  options: ISpecTestOptions<TestCase, Result>
```
Comment (wemeetagain) [inline packages/eth2.0-spec-test-util/src/single.ts]: Maybe makes sense to have a `Partial` here

File: packages/eth2.0-spec-test-util/src/single.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+      expect(testFunction.bind(null, testCase)).to.throw;
+    } else {
+      const profileId = `${name}-${Date.now()}.profile`;
+      if (process.env.GEN_PROFILE_DIR) {
+        profiler.startProfiling(profileId);
+      }
+      const result = testFunction(testCase);
+      if (process.env.GEN_PROFILE_DIR) {
+        const profile = profiler.stopProfiling(profileId);
+        const directory = process.env.GEN_PROFILE_DIR || __dirname;
```
Comment (wemeetagain) [inline packages/eth2.0-spec-test-util/src/single.ts]: I don't think we want __dirname here. We can pull out the declaration of profilingDirectory, and use that for the `if`s.

**Review by wemeetagain — CHANGES_REQUESTED** [review]
Looks really good. I like the `ISpecTestOptions`, it will be much easier to configure.

File: packages/eth2.0-spec-test-util/src/single.ts (hunk 3 above)
Comment (mpetrunic) [reply packages/eth2.0-spec-test-util/src/single.ts]: dirname is just defaul value if env is not provided. I dont understand your solution though

Comment (wemeetagain) [reply packages/eth2.0-spec-test-util/src/single.ts]: we already know `process.env.GEN_PROFILE_DIR` is set, bc of L112.
We could instead, after L107 do: `const profilingDir = process.env.GEN_PROFILE_DIR`, then `if(profilingDir) doProfile`

Comment (mpetrunic) [reply packages/eth2.0-spec-test-util/src/single.ts]: oh yeah, brain fart haha

**Review by wemeetagain — APPROVED** [review]

**Review by GregTheGreek — APPROVED** [review]

**Pull request merged**